source/tutorials/source-file-selection.md

```diff
@@ -0,0 +1,645 @@
+  # To nicely show path contents
+  nativeBuildInputs = [ tree ];
+} ''
+  tree ${./.}
+''
+```
+
+Running it gives us:
+```shell-session
+$ nix-build -A package
+building '/nix/store/6ybg4v48xy8azhrnfdccdmhd2gr938f5-directory-interpolation.drv'...
+/nix/store/xdfchqpfx20ar9jil9kys99wc6hnm9zx-select
+|-- default.nix
+|-- nix
+|   |-- sources.json
+|   `-- sources.nix
+|-- package.nix
+`-- string.txt
+```
+
+It's also very common to use this for the [`src`](https://nixos.org/manual/nixpkgs/stable/#var-stdenv-src) variable in `stdenv.mkDerivation` like so:
+
+```nix
+{ stdenv, tree }:
+stdenv.mkDerivation {
+  name = "directory-coercion";
+  src = ./.;
+  nativeBuildInputs = [ tree ];
+  postInstall = ''
+    touch created-in-the-build
+    tree
+  '';
+}
+```
+
+Setting `src` will copy the resulting store path into the build directory and mark it as mutable.
+For the many commands that expect to be able to write to the current directory, this is great:
+
+```shell-session
+$ nix-build -A package
+building '/nix/store/2cqd93fpnb4vqwkwmbl66dbxhndq1vhh-directory-coercion.drv'...
+unpacking sources
+unpacking source archive /nix/store/178fbwa8iwdl6b85yafksdbwlxf6mjca-select
+[...]
+.
+|-- created-in-the-build
+|-- default.nix
+|-- nix
+|   |-- sources.json
+|   `-- sources.nix
+|-- package.nix
+`-- string.txt
+```
+
+However there are some subtle problems with this approach:
+- Note how the name of the store path ends with `-select`.
+  So the name of the local directory influenced the result.
+
+  This means that whenever you rename the project directory
+  or a collegue runs it in a different directory name,
+  you're going to get different build results!
+
+- All files in the directory are unconditionally added to the Nix store.
+
+  This means that:
+  - Even if your build only needs a few files,
+    changing _any_ file in the directory requires rebuilding the derivation,
+    potentially wasting a lot of time.
+
+  - If you have any secrets in the current directory,
+    they get imported into the Nix store too, exposing them to all users on the system!
+
+### `builtins.path`
+
+<!-- TODO: Use lib.cleanSourceWith instead and only briefly mention builtins.path? -->
+
+The above problems can be fixed by using [`builtins.path`](https://nixos.org/manual/nix/stable/language/builtins.html#builtins-path) instead.
+It allows customising the name of the resulting store path with its `name` argument.
+And it allows selecting the files that should be included with its `filter` argument.
+
+```nix
+{ runCommand, tree, lib }:
+let
+  src = builtins.path {
+    # The convention is to use "source"
+    name = "source";
+    path = ./.;
+
+    filter = pathString: type:
+      # Recurse into directories
+      type == "directory"
+      # Don't include .nix files
+      || ! lib.hasSuffix ".nix" pathString;
+  };
+in
+stdenv.mkDerivation {
+  name = "builtins-path";
+  inherit src;
+  nativeBuildInputs = [ tree ];
+  postInstall = ''
+    tree
```

Comment:
I am not sure why the `touch created-in-the-build` was in the previous example, but omitted in this one - maybe it's not important.